pkg/api/metadata.go

```diff
@@ -0,0 +1,197 @@
+/*
+Copyright 2021 GramLabs, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package api
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"reflect"
+	"strings"
+	"time"
+)
+
+const (
+	RelationSelf      = "self"
+	RelationNext      = "next"
+	RelationPrev      = "prev"
+	RelationAlternate = "alternate"
+	RelationLabels    = "https://stormforge.io/rel/labels"
+	RelationTrials    = "https://stormforge.io/rel/trials"
+	RelationNextTrial = "https://stormforge.io/rel/next-trial"
+)
+
+// Metadata is used to hold single or multi-value metadata from list responses.
+type Metadata map[string][]string
+
+func (m Metadata) Title() string {
+	return http.Header(m).Get("Title")
+}
+func (m Metadata) SetTitle(value string) {
+	if value != "" {
+		http.Header(m).Set("Title", value)
+	}
+}
+
+func (m Metadata) Location() string {
+	return http.Header(m).Get("Location")
+}
+func (m Metadata) SetLocation(value string) {
+	if value != "" {
+		http.Header(m).Set("Location", value)
+	}
+}
+
+func (m Metadata) LastModified() time.Time {
+	value, _ := http.ParseTime(http.Header(m).Get("Last-Modified"))
+	return value
+}
+func (m Metadata) SetLastModified(t time.Time) {
+	if !t.IsZero() {
+		http.Header(m).Set("Last-Modified", t.UTC().Format(http.TimeFormat))
+	}
+}
+
+func (m Metadata) Link(rel string) string {
+	for _, rh := range http.Header(m).Values("Link") {
+		for _, h := range strings.Split(rh, ",") {
+			r, l := splitLink(h)
+			if strings.EqualFold(rel, r) {
+				return l
+			}
+		}
+	}
+	return ""
+}
+func (m Metadata) SetLink(rel, link string) {
```

Comment:
Do we want this to handle injecting the base url? Perhaps this means the const Relations would change into paths instead of the full url?

Reply:
The link relations are just opaque strings, the standard relations (e.g. "next", "prev") won't use the URI form, while our "custom" relations will. I'd rather not try anything fancy so we have the full value of the link relations grep-able.

It also just occurred to me that, for the most part, we will just be consuming links (i.e. the `Link` header is a _response_ header not a _request_ header). I would be equally happy to just remove the `Set` functions altogether.

Reply:
I removed the `Set*` functions since they aren't actually being used.